lib/lf/uscalfe/scalar_fe_space.h

```diff
@@ -0,0 +1,147 @@
+ *
+ * @tparam SCALAR underlying scalar type, usually either `double` or
+ * `complex<double>`
+ *
+ * The abstract concept of a (parametric) finite element space involves
+ * - an underlying mesh
+ * - the definition of a local set of shape functions for every cell.
+ *
+ * This class just contains (pointers to) objects representing the various
+ * building blocks of a finite element space. It does not offer elaborate
+ * methods.
+ *
+ * @note Some of the pointers may be NULL. For instance, if all computations
+ *       are done on purely triangular meshes then a finite element
+ * specification for quadrilaterals need not be given.
+ *
+ * This class is covered in @\lref{par:fespace}.
+ */
+template <typename SCALAR>
+class ScalarFESpace {
+ public:
+  using Scalar = SCALAR;
+
+  /** @brief default constructor, needed by std::vector
+   * @note creates an invalid object that cannot be used. */
+  ScalarFESpace() = default;
+  ScalarFESpace(const ScalarFESpace &) = delete;
+  ScalarFESpace(ScalarFESpace &&) noexcept = default;
+  ScalarFESpace &operator=(const ScalarFESpace &) = delete;
+  ScalarFESpace &operator=(ScalarFESpace &&) noexcept = default;
+  /**
+   * @brief Main constructor: sets up the local-to-global index mapping (dof
+   * handler)
+   *
+   * @param mesh_p shared pointer to underlying mesh (immutable)
+   */
+  ScalarFESpace(std::shared_ptr<const lf::mesh::Mesh> mesh_p)
+      : mesh_p_(std::move(mesh_p)) { }
+
+  /** @brief acess to underlying mesh
+   *  @return a shared _pointer_ to the mesh
+   */
+  [[nodiscard]] std::shared_ptr<const lf::mesh::Mesh> Mesh() const {
+    LF_VERIFY_MSG(mesh_p_ != nullptr, "No valid FE space object: no mesh");
+    return mesh_p_;
+  }
+  /** @brief access to associated local-to-global map
+   * @return a reference to the lf::assemble::DofHandler object (immutable)
+   */
+  [[nodiscard]] virtual const lf::assemble::DofHandler &LocGlobMap() const = 0;
+
+  /** @brief access to shape function layout for cells
+   *
+   * @param entity The entity to get the reference element for
+   *
+   * @warning NULL pointers may be returned by this method in case a finite
+   * element specification was not given for a particular topological type of
+   * entity.
+   */
+  [[nodiscard]] virtual std::shared_ptr<const ScalarReferenceFiniteElement<SCALAR>>
+  ShapeFunctionLayout(const lf::mesh::Entity &entity) const = 0;
+
+  /** @brief number of _interior_ shape functions associated to entities of
+   * various types
+   */
+  [[nodiscard]] virtual size_type NumRefShapeFunctions(const lf::mesh::Entity &entity) const = 0;
+
+  /** @brief No special destructor */
+  virtual ~ScalarFESpace() = default;
+
+ private:
+  /** Underlying mesh */
+  std::shared_ptr<const lf::mesh::Mesh> mesh_p_;
+
+ public:
+  /** Output control variable */
+  static unsigned int ctrl_;
+  static const unsigned int kout_mesh = 1;
+  static const unsigned int kout_dofh = 2;
+  static const unsigned int kout_rsfs = 4;
```

Comment:
These variables come from the original implementation of `lf::uscalfe::UniformScalarFESpace` which was implemented by @hiptmair. So it would probably be best if he could take care of their documentation.

Reply:
I will do this in the end.